web/packages/teleterm/src/ui/hooks/usePersistedState.ts

```diff
@@ -94,16 +94,25 @@ export function usePersistedState<
   // listener in statePersistenceService that gets called whenever the given key gets updated.
   const [, rerender] = useState<object>();
 
-  const setState = useCallback(
-    (newState: WholeState[Key]) => {
-      statePersistenceService.putState({
-        ...statePersistenceService.getState(),
-        [key]: newState,
-      });
+  const setState: Dispatch<SetStateAction<WholeState[Key]>> = useCallback(
+    newState => {
+      if (typeof newState === 'function') {
```

Comment:
The type for `Dispatch<SetStateAction<Foo>>` is weird like that, maybe I'm doing something wrong.  I'm basing this on [`useStateRef`](https://github.com/gravitational/teleport/blob/c33a85bf9e40e385d049d88d44ebb9825e312bc4/web/packages/shared/hooks/useStateRef.ts#L53-L70) where I've also used a similar pattern.

Reply:
I actually meant to leave this comment on line 102, because having to cast `newState` is the weird part. Checking `typeof newState === 'functcion'` for some reason doesn't narrow it down to a function only.

Reply:
Yeah, it's weird. I don't understand why the check doesn't work here.